book/html.md

```diff
@@ -142,66 +224,130 @@ sibling to [troff][troff], now used for Linux man pages. The
+means defining the `__repr__` function:
 
 ``` {.python expected=False}
-def parse(tokens):
-    # ...
-    print(currently_open)
-    raise Exception("Reached last token before end of document")
+class Text:
+    def __repr__(self):
+        return repr(self.text)
+
+class Element:
+    def __repr__(self):
+        return "<" + self.tag + ">"
 ```
 
-Python prints a list of `ElementNode` objects, meaning that there were
-open HTML elements still around when it reached the last token:
+Try this out on this web page, parsing the HTML source code and then
+calling `print_tree` to visualize it:
 
+``` {.python expected=False}
+headers, body = request(sys.argv[1])
+nodes = HTMLParser(body).parse()
+print_tree(nodes)
 ```
-[<__main__.ElementNode object at 0x101399c70>,  ...]
+
+Run it on this web page, and you'll see something like this:
+
+``` {.example}
+ <!DOCTYPE html >
+   '\n'
+   <html lang="en-US" xml:lang="en-US" >
+     '\n'
+     <head >
+       '\n  '
+       <meta charset="utf-8" / >
+         '\n  '
+         <meta name="generator" content="pandoc" / >
+           '\n  '
 ```
 
-Ok, that's not too helpful. Python needs a method called `__repr__` to
-be defined to print things a little more reasonably:
+Immediately a couple of things stand out. Let's start at the top, with the `<!DOCTYPE html>` tag.
 
-``` {.python}
-class ElementNode:
+This special tag is called a [doctype][html5-doctype] that's s always the very first thing in an HTML document. But it's not really an element at all, nor is it supposed to have close tag. Our toy browser won't be using the doctype for anything, so it's best to throw it away:[^quirks-mode]
+
+[html5-doctype]: https://html.spec.whatwg.org/multipage/syntax.html#the-doctype
+
+[^quirks-mode]: Real browsers use doctypes to switch between
+    standards-compliant and legacy parsing and layout modes.
+
+``` {.python indent=4}
+def add_tag(self, tag):
+    if tag.startswith("!"): return
     # ...
-    def __repr__(self):
-        return "<" + self.tag + ">"
 ```
 
-This produces a more reasonable result:
+This ignores all tags that start with an exclamation mark, which not only throws out doctype declarations but also most comments, which in HTML are written `<!-- comment text -->`.
 
-``` 
-[<!DOCTYPE html>,
- <html lang="en-US" xml:lang="en-US">,
- <head>,
- <meta charset="utf-8" />,
- <meta name="generator" content="pandoc" />,
- <meta name="viewport" content=... />,
- <link rel="prev" href="text" />,
- <link rel="next" href="layout" />,
- <link rel="stylesheet" href="../book.css" />]
-```
+Just throwing out doctypes isn't quite enough though---if you run your parser now, it will crash. That's because after the doctype comes a newline, which our parser treats as text and tries to insert into the tree. Except there isn't a tree, since there haven't been any open tags. For simplicity, let's just have our browser skip whitespace-only text nodes to side-step the problem:[^ignore-them]
```

Comment:
```suggestion
Just throwing out doctypes isn't quite enough though--if you run your parser now, it will crash. That's because after the doctype comes a newline, which our parser treats as text and tries to insert into the tree. Except there isn't a tree, since there haven't been any open tags. For simplicity, let's just have our browser skip whitespace-only text nodes to side-step the problem:[^ignore-them]
```

Reply:
This one is correct, actually. `pandoc` converts triple-dash to an em dashes (for pauses in sentences) and double-dash to en dashes (for ranges).